### ISSUE TITLE
Provide better error message when IJulia is installed somewhere other than the global environment

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/IJulia.jl#1127]
Currently, when one installs IJulia in a temporary environment (e.g. to test out the package) the kernel fails to connect, with no obvious error message. This led me to raise issue #1126. Ideally, in this case there would be an error message or warning of some sort indicating what the problem is.

**Comment by LilithHafner** [comment]
Even better would be to not require that IJulia is installed in the global environment.

**Comment by JamesWrigley** [comment]
After thinking about this a bit, I feel that the best implementation would be to get Jupyter itself to show an error if launching the kernel failed. Otherwise we would basically need to implement a mini-kernel in `kernel.jl` that implements enough of the kernel spec to report an error.

**Comment by lewisl** [comment]
That would be nice but no one working on Julia controls or even works on Jupyter.  We need a Julia solution.  Perhaps just some documentation about how to get project focused installs to work.  Seems like the kernelspec should ALWAYS work and project specific stuff goes into a project and into notebooks created within those projects.

**Comment by goerz** [comment]
I was able to confirm this, but only if the kernel spec does not include `--project=@.`. That is, in addition to the instructions in #1126, I had to also run `IJulia.installkernel("Julia Test")` to see the failure. If, instead, I do `IJulia.installkernel("Julia Test", "--project=@.")`, the kernel will connect. I also had to make sure to uninstall `IJulia` from the base environment to get it to fail.

If I run `jupyter lab` from the command line, instead of using `IJulia.notebook()` from the Julia REPL, `jupyter` shows an appropriate error message in the terminal:

```
ERROR: LoadError: ArgumentError: Package IJulia not found in current path.
- Run `import Pkg; Pkg.add("IJulia")` to install the IJulia package.
```

This makes perfect sense, if you look at the kernel spec that was generated, and would look something like this:

```
{
  "display_name": "Julia Test 1.11.5",
  "argv": [
    "/Users/goerz/.julia/juliaup/julia-1.11.5+0.aarch64.apple.darwin14/bin/julia",
    "-i",
    "--color=yes",
    "/Users/goerz/.julia/packages/IJulia/XF6bn/src/kernel.jl",
    "{connection_file}"
  ],
  "language": "julia",
  "env": {},
  "interrupt_mode": "signal"
}
```

With those options, that will start `julia` with the base environment, and then try to run `.../src/kernel.jl`, which imports `IJulia`. That has to fail if `IJulia` is not in the base environment. As soon as you add the `--project=@.` option to that, it works, because then `IJulia` _is_ in the environment that the kernel runs in.

I don't really see a way around this. Maybe `IJulia` can detect _during package installation_ whether the current environment is a base environment and show a big warning if it not. It's a bit unfortunate that `IJulia.notebook()` doesn't just show the output from `jupyter lab`, and thus eats the error message. Maybe that's something that can be improved. (Although, personally, I consider `IJulia.notebook()` a misfeature: the IJulia package should focus on providing a kernel, not on managing a Jupyter installation. Arguably, _Julia_ has no business managing a Jupyter installation. Maybe for the convenience of people who lack the skill to install Jupyter, but then that should probably be a `Jupyter` package that is independent of `IJulia`. Anyway, I digress. Rant off)

So, _either_ of the following two things solve the issue:

- Make sure `IJulia` is installed in the base environment for the `julia` executable referenced in the kernelspec (possibly in _addition_ to the local/temporary environment)
- Make sure the `kernelspec` contains `--project=@.`, either by using the default kernel or by manually running `IJulia.installkernel` with that option.

Arguably, "don't use `IJulia.notebook()`, but run `jupyter lab` from the terminal" also solves the issue, as it makes it quite hard to run into this situation in the first place, and gives a meaningful error message when it does ;-)

**Comment by lewisl** [comment]
I almost always run Jupyter from the command line as jupyter lab or jupyter notebook. The only time I do not is when I must do a fresh install of Julia packages for a new version of Julia (I’d endorse juliaup heartily if it could do this!) and I want to test my install of IJulia.

On May 12, 2025, at 10:23 AM, Michael Goerz ***@***.***> wrote:

[https://avatars.githubusercontent.com/u/112306?s=20&v=4]goerz left a comment (JuliaLang/IJulia.jl#1127)<https://github.com/JuliaLang/IJulia.jl/issues/1127#issuecomment-2873373304>

I was able to confirm this, but only if the kernel spec does not include ***@***.*** That is, in addition to the instructions in #1126<https://github.com/JuliaLang/IJulia.jl/issues/1126>, I had to also run IJulia.installkernel("Julia Test") to see the failure. If, instead, I do IJulia.installkernel("Julia Test", ***@***.***"), the kernel will connect. I also had to make sure to uninstall IJulia from the base environment to get it to fail.

If I run jupyter lab from the command line, instead of using IJulia.notebook() from the Julia REPL, jupyter shows an appropriate error message in the terminal:

ERROR: LoadError: ArgumentError: Package IJulia not found in current path.
- Run `import Pkg; Pkg.add("IJulia")` to install the IJulia package.


This makes perfect sense, if you look at the kernel spec that was generated, and would look something like this:

{
  "display_name": "Julia Test 1.11.5",
  "argv": [
    "/Users/goerz/.julia/juliaup/julia-1.11.5+0.aarch64.apple.darwin14/bin/julia",
    "-i",
    "--color=yes",
    "/Users/goerz/.julia/packages/IJulia/XF6bn/src/kernel.jl",
    "{connection_file}"
  ],
  "language": "julia",
  "env": {},
  "interrupt_mode": "signal"
}


With those options, that will start julia with the base environment, and then try to run .../src/kernel.jl, which imports IJulia. That has to fail if IJulia is not in the base environment. As soon as you add the ***@***.*** option to that, it works, because then IJulia is in the environment that the kernel runs in.

I don't really see a way around this. Maybe IJulia can detect during package installation whether the current environment is a base environment and show a big warning if it not. It's a bit unfortunate that IJulia.notebook() doesn't just show the output from jupyter lab, and thus eats the error message. Maybe that's something that can be improved. (Although, personally, I consider IJulia.notebook() a misfeature: the IJulia package should focus on providing a kernel, not on managing a Jupyter installation. Arguably, Julia has no business managing a Jupyter installation. Maybe for the convenience of people who lack the skill to install Jupyter, but then that should probably be a Jupyter package that is independent of IJulia. Anyway, I digress. Rant off)

So, either of the following two things solve the issue:

  *   Make sure IJulia is installed in the base environment for the julia executable referenced in the kernelspec (possibly in addition to the local/temporary environment)
  *   Make sure the kernelspec contains ***@***.***, either by using the default kernel or by manually running IJulia.installkernel with that option.

Arguably, "don't use IJulia.notebook(), but run jupyter lab from the terminal" also solves the issue, as it makes it quite hard to run into this situation in the first place, and gives a meaningful error message when it does ;-)

**Comment by goerz** [comment]
Actually, I have to correct myself slightly: `--project=@.` would not have solved the exact problem in #1126. I got confused because while trying to reproduce the issue, I also instantiated a project in my test folder. The more general point is: whatever environment the command in the kernelspec runs in has to have `IJulia` installed. With `pkg> activate --temp`, that would have to be `--project=<absolute path to temp environment>`. To be fair, that's quite annoying for a temporary environment, since it's tricky to get that path, and you'd have to manually delete the kernel afterwards. So, practically speaking: Always make sure you have `IJulia` installed in the base environment, or some other shared environment, as [James prefers](https://github.com/JuliaLang/IJulia.jl/issues/1127#issuecomment-2476112600).

> Otherwise we would basically need to implement a mini-kernel in kernel.jl that implements enough of the kernel spec to report an error.

I'm not sure it has to go quite that far: what if instead of executing `kernel.jl`, the kernelspec referenced a script that only used stdlib functions (and, specifically, does not import `IJulia`). That script could then perform some checks, ensures that there's an environment with `IJulia` available, and then does `exec` for a subprocess running what the _current_ `kernel.jl` is. That would at least give a cleaner error message, although I'm not sure it would help if people run `IJulia.notebook()`. In either case, the kernel "dies" and the error message that Jupyter shows seems to get eaten.

**Comment by goerz** [comment]
@lewisl Is there any chance that you can go on GitHub and clean up the formatting in https://github.com/JuliaLang/IJulia.jl/issues/1127#issuecomment-2873400790 and comments in #1152? The "reply by mail" that you're using produces unreadable posts, in addition to [garbling the text](https://github.com/JuliaLang/IJulia.jl/issues/1152#issuecomment-2870061743). Not entirely your fault: [GitHub should be ashamed of themselves](https://github.com/orgs/community/discussions/77491) for not processing emails like that better. Alternatively, you would probably get okay results if you deleted the texts of the original comments you're responding to. That is, make sure your email contains _only_ the text you want to be in the message.

**Comment by lewisl** [comment]
OK.  1127 seems ok except for inline message which I removed.

1152 I removed inline message.  other than being too long the comments looked ok.

did you also want me to remove the reply to links that github puts in?

**Comment by lewisl** [comment]
Yes, of course. It’s old.  I can do an activate . before I launch.  Haven’t seen the same problem since.

On Sep 7, 2025, at 9:22 AM, James Wrigley ***@***.***> wrote:

[https://avatars.githubusercontent.com/u/5361518?s=20&v=4]JamesWrigley left a comment (JuliaLang/IJulia.jl#1127)<https://github.com/JuliaLang/IJulia.jl/issues/1127#issuecomment-3263773235>

I'm going to close this because I don't think there's anything we can do from IJulia. FWIW, at least now jupyterlab() and notebook() support passing verbose=true so you can see any errors from Jupyter: https://julialang.github.io/IJulia.jl/stable/manual/troubleshooting/#Kernel-connection-failure-tips
